cmake/ccache.cmake

```diff
@@ -0,0 +1,155 @@
+
+  ``QUIET`` Silent output from this method.
+
+#]=======================================================================]
+function(enable_ccache)
+    set(options QUIET)
+    set(oneValueArgs)
+    set(multiValueArgs LANGUAGES OPTIONS)
+    cmake_parse_arguments(CCACHE "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN} )
+
+    if (CCACHE_UNPARSED_ARGUMENTS)
+        message(FATAL_ERROR "Morpheus: enable_ccache invalid arguments ${CCACHE_UNPARSED_ARGUMENTS}")
+    endif()
+    
+    find_program(CCACHE_BIN ccache)
+    if (NOT CCACHE_BIN)
+        message(STATUS "Morpheus: CCache not found. Rebuilding all the source files.")
+        message(DEBUG "Morpheus: If CCache compiler cache is desired then ensure this is a supported platform and ensure the Conan buildenv is active")
+        return()
+    endif()
+    message(STATUS "Morpheus: CCache found: ${CCACHE_BIN}. Enabling ccache as compiler cache tool.")
+
+    if (CCACHE_DEBUG)
+        message(STATUS "Morpheus: Create CCache debug files")
+        list(APPEND ccacheEnv "CCACHE_DEBUG=1")
+    endif()
+    if (CCACHE_ICECC)
+        find_program(ICECC_BIN icecc)
+        if (ICECC_BIN)
+            message(STATUS "Morpheus: build with Icecream ${ICECC_BIN}")
+            list(APPEND ccacheEnv "CCACHE_PREFIX=icecc")
+        endif()
+    endif()
+    if (CCACHE_SLOPPINESS)
+        message(STATUS "Morpheus: Enable CCache sloppiness")
+        list(APPEND ccacheEnv "CCACHE_SLOPPINESS=pch_defines,time_macros")
+    endif()
+    string(REPLACE ";" " " ccacheEnvString "${ccacheEnv}")
+
+    if (${CMAKE_GENERATOR} MATCHES "Ninja|Makefiles" AND
+        ${CMAKE_HOST_SYSTEM_NAME} MATCHES "Linux")
+        message(STATUS "Morpheus: found generator ${CMAKE_GENERATOR} and OS ${CMAKE_HOST_SYSTEM_NAME}")
+        foreach(lang IN ITEMS ${LANGUAGES})
+            set(CMAKE_${lang}_COMPILER_LAUNCHER
+                ${CMAKE_COMMAND} -E env ${ccacheEnvString} ${CCACHE_BIN})
+        endforeach()
+    elseif (${CMAKE_GENERATOR} STREQUAL "Xcode")
+        message(STATUS "Morpheus: found generator ${CMAKE_GENERATOR}")
+        foreach(lang IN ITEMS ${LANGUAGES})
+            set(launch${lang} ${CMAKE_BINARY_DIR}/launch-${lang})
+            file(WRITE ${launch${lang}} "#!/bin/bash\n\n")
+            foreach(envVal IN ITEMS ${ccacheEnv})
+                file(APPEND ${launch${lang}} "export ${envVal}\n")
+            endforeach() 
+            file(APPEND ${launch${lang}} "exec \"${CCACHE_BIN}\" \"${CMAKE_${lang}_COMPILER}\" \"$@\"\n")     
+            execute_process(COMMAND chmod a+rx ${launch${lang}})
+            set(${CMAKE_XCODE_ATTRIBUTE_${lang}} ${launch${lang}})
+            if(lang STREQUAL "C")
+                set(CMAKE_XCODE_ATTRIBUTE_LD ${launchC}) 
+            elseif(lang STREQUAL "CXX")
+                set(CMAKE_XCODE_ATTRIBUTE_LDPLUSPLUS ${launchCXX})
+            endif()
+        endforeach()
+    elseif (${CMAKE_GENERATOR} MATCHES "Visual Studio|Ninja" AND
```

Comment:
Does this check for Ninja ever get used given we've already checked for Ninja above?  Does this need to be some way of detecting if its Ninja used with Visual Studio?

Reply:
Yes, you are absolutely right. l have it in the to-do list to clean up repetitions and optimize the code. Let's say that for the moment I am trying to get a first version of the functionality working for all the generators as fast as possible. I read somewhere that as a professional approach of coding, your inital goal must be to have the functionality working as fast as possible. Then after the functionality is working you invest time in cleaning up and optimizing the code. According to your experience, what do you think about this approach? Is it correct or it's better to spend a little more time at the beginning of the development and release a more refined first version of the code with a better quality product?

Reply:
Sounds good.  I'm definitely for getting something working so that we can iterate on that and get feedback from trying it out.  With something like this part of what we want to do is build our experience of CCache so we understand how it works, and thus can tie it in correctly.

While we are thinking about this we should think about how we might test this.  I've worked with CCache before, and if something breaks it then it can require manual investigation to confirm it works.  Perhaps we should create a github action workflow that performs a build (we can test all relevant configurations then, i.e. Ninja, Makefile, XCode, Visual Studio + MSBuild, VS + Ninja).  We can then query Ccache at the beginning and end to ensure that files are added to the cache, we can then rebuild (perhaps not the whole project, but just a single target) to ensure that files are retrieved from the cache.  We'd need to write something to capture the CCache output and extract the data we want to confirm it's working.  Does this sound reasonable?

Reply:
I agree. I will proceed in this way, thanks!

Reply:
With the last PR I was able to successfully build with CCache with generators Ninja Multi-Config, Makefiles and Visual Studio. The hardest one was getting MSBuild work. There are known issue with it and CMake integration: https://gitlab.kitware.com/cmake/cmake/-/issues/24864
I am now proceeding to change all the workflows in order to add a functionality to show what CCache is doing.
Some resources here should hopefully help to do not reinvent the wheel from scratch:
https://cristianadam.eu/20200113/speeding-up-c-plus-plus-github-actions-using-ccache/
https://github.com/hendrikmuhs/ccache-action
https://github.com/canonical/multipass/blob/9056c2cdb47677732dd0a0e2a4dd6d7385a0a69a/.github/workflows/linux.yml